### PR TITLE
Replace `repo.namedBlob` by `git.TreeEntry`.

### PR DESCRIPTION
**Pull request by kousu**
`namedBlob` turned out to be a poor imitation of a `TreeEntry`. Using the latter directly shortens this code.

This partially undoes https://github.com/go-gitea/gitea/pull/23152/, which I found a merge conflict with, and also expands the test it added to cover the subtle README-in-a-subfolder case.

### CODE REVIEW

**Review by delvh — APPROVED** [review]
I just asked myself if this cache type was implemented for performance reasons, but according to the method implementations, that is negligible…
Was that different previously?
Otherwise, I don't know how this could happen.

**Comment by kousu** [comment]
> for performance reasons

And even if it was, it was only optimizing for README files. How heavy is that really?

>  Otherwise, I don't know how this could happen.

Thanks for the vote of confidence!

**Comment by zeripath** [comment]
IIRC we needed the namedBlob at some point because we were passing round git.Blobs but we're able to get the TreeEntry just as easily now so it might be needed anymore

File: routers/web/repo/view.go [hunk 1]
```diff
@@ -98,28 +92,22 @@ func findReadmeFileInEntries(ctx *context.Context, entries []*git.TreeEntry) (*n
...
+						return "", nil, err
 					}
 				}
 				if target != nil && (target.IsExecutable() || target.IsRegular()) {
-					readmeFiles[i] = &namedBlob{
-						name,
-						isSymlink,
-						target.Blob(),
-					}
+					readmeFiles[i] = target
```
Comment (lunny) [inline routers/web/repo/view.go]: The logic here has been changed. The previous name comes from the original file, not the target file.

Comment (kousu) [reply routers/web/repo/view.go]: Oh I see. That's subtle. With a symlink in the repo like this:

```
sdffsf$ tree .github/
.github/
├── a
│   └── b
│       └── c
│           └── _hocs.md
└── README.md -> a/b/c/_hocs.md

3 directories, 2 files
```

6976d703d5776e556b8644082468aff2a51f484a renders with the nonexistent path ".github/_hocs.md"

Screenshot 2023-02-13 at 22-38-33 sdffsf

I've moved (d9393ee3bb1ae5f8fbc9225dc71e7479195a6ae4) the symlink checking logic into renderReadmeFile() and now it renders like it used to:

Screenshot 2023-02-13 at 22-37-32 sdffsf

File: routers/web/repo/view.go [hunk 2]
```diff
@@ -98,28 +92,12 @@ func findReadmeFileInEntries(ctx *context.Context, entries []*git.TreeEntry) (*n
...
-				}
-				if target != nil && (target.IsExecutable() || target.IsRegular()) {
-					readmeFiles[i] = &namedBlob{
-						name,
-						isSymlink,
-						target.Blob(),
-					}
-				}
+			if readmeFiles[i] == nil || base.NaturalSortLess(readmeFiles[i].Name(), entry.Blob().Name()) {
+				readmeFiles[i] = entry
```
Comment (lunny) [inline routers/web/repo/view.go]: I think we still need to check if the readme file is valid here. Assume we have both a broken link readme file and a valid readme file in different directories or names. Moving the logic to `renderReadmeFile` will result in no README file will be displayed.

Comment (kousu) [reply routers/web/repo/view.go]: Okay, you're right. I'm going to table this PR and write tests first (like I did for https://github.com/go-gitea/gitea/pull/22196) and then we'll be able to know for sure if this is behaving.

**Comment by kousu** [comment]
drafted until there are tests pre-written to cover README rendering: https://github.com/go-gitea/gitea/pull/22898#discussion_r1110009347

File: routers/web/repo/view.go (hunk 2 above)
Comment (kousu) [reply routers/web/repo/view.go]: Tests are in https://github.com/go-gitea/gitea/pull/23185; the test for this specific case is at 

https://github.com/go-gitea/gitea/blob/8afde8142010985c6d3b6b98eb0be96d7b24da52/tests/integration/repo_test.go#L335-L336

I'll rebase and see if my code passes.

**Comment by kousu** [comment]
The [current CI failures](https://drone.gitea.io/go-gitea/gitea/69315/) are

```
Init storage failed: Put "http://minio:9000/gitea/": dial tcp: lookup minio on 127.0.0.11:53: no such host
```

and

```
=== Test_DeleteOrphanedIssueLabels (/drone/src/models/migrations/v1_14/v177_test.go:37)

LoadFixtures failed after retries: testfixtures: error inserting record: pq: column "exclusive" of relation "label" does not exist, on file: label.yml, index: 0, sql: INSERT INTO "label" ("num_closed_issues", "id", "repo_id", "org_id", "name", "color", "exclusive", "num_issues") VALUES ($1, $2, $3, $4, $5, $6, $7, $8), params: [0 1 1 0 label1 #abcdef false 2]

--- FAIL: Test_DeleteOrphanedIssueLabels (5.07s)

    v177_test.go:37: initializing fixtures from: /drone/src/models/migrations/fixtures/Test_DeleteOrphanedIssueLabels

    v177_test.go:37: error whilst loading fixtures from /drone/src/models/migrations/fixtures/Test_DeleteOrphanedIssueLabels: testfixtures: error inserting record: pq: column "exclusive" of relation "label" does not exist, on file: label.yml, index: 0, sql: INSERT INTO "label" ("num_closed_issues", "id", "repo_id", "org_id", "name", "color", "exclusive", "num_issues") VALUES ($1, $2, $3, $4, $5, $6, $7, $8), params: [0 1 1 0 label1 #abcdef false 2]

FAIL

FAIL	code.gitea.io/gitea/models/migrations/v1_14	6.460s

FAIL
```

These don't seem to be related to this patch. I'll restart and if it happens again I guess I'll try to reproduce locally.

File: models/fixtures/repository.yml [hunk 3]
```diff
@@ -25,7 +25,7 @@
   fork_id: 0
   is_template: false
   template_id: 0
-  size: 7028
+  size: 7320
```
Comment (kousu) [inline models/fixtures/repository.yml:28]: This had to change due to the test additions I did. I used the same test repo (`user2/repo1.git`) that https://github.com/go-gitea/gitea/pull/23152 did, but I don't know why that PR didn't force changing this number.

File: routers/web/repo/view.go [hunk 4]
```diff
@@ -306,19 +300,19 @@ func renderReadmeFile(ctx *context.Context, readmeFile *namedBlob, readmeTreelin
...
-			RelativePath: path.Join(ctx.Repo.TreePath, readmeFile.name), // ctx.Repo.TreePath is the directory not the Readme so we must append the Readme filename (and path).
-			URLPrefix:    path.Dir(readmeTreelink),
+			RelativePath: path.Join(ctx.Repo.TreePath, readmeFile.Name()), // ctx.Repo.TreePath is the directory not the Readme so we must append the Readme filename (and path).
+			URLPrefix:    path.Join(readmeTreelink, subfolder),
 			Metas:        ctx.Repo.Repository.ComposeDocumentMetas(),
 			GitRepo:      ctx.Repo.GitRepo,
 		}, rd)
 		if err != nil {
-			log.Error("Render failed for %s in %-v: %v Falling back to rendering source", readmeFile.name, ctx.Repo.Repository, err)
+			log.Error("Render failed for %s in %-v: %v Falling back to rendering source", readmeFile.Name(), ctx.Repo.Repository, err)
```
Comment (kousu) [inline routers/web/repo/view.go:315]: Perhaps

```suggestion
			log.Error("Render failed for %s in %-v: %v Falling back to rendering source", path.Join(subfolder, readmeFile.Name()), ctx.Repo.Repository, err)
```
?

File: routers/web/repo/view.go (hunk 2 above)
Comment (kousu) [reply routers/web/repo/view.go]: Passing in https://drone.gitea.io/go-gitea/gitea/69318/3/10#L848 !

File: routers/web/repo/view.go [hunk 5]
```diff
@@ -270,25 +254,35 @@ func getFileReader(repoID int64, blob *git.Blob) ([]byte, io.ReadCloser, *fileIn
...
+	buf, dataRc, fInfo, err := getFileReader(ctx.Repo.Repository.ID, target.Blob())
 	if err != nil {
 		ctx.ServerError("getFileReader", err)
 		return
 	}
 	defer dataRc.Close()
 
 	ctx.Data["FileIsText"] = fInfo.isTextFile
-	ctx.Data["FileName"] = readmeFile.name
+	ctx.Data["FileName"] = path.Join(subfolder, readmeFile.Name())
```
Comment (lunny) [inline routers/web/repo/view.go:281]: should we display the link name or target name?

Comment (kousu) [reply routers/web/repo/view.go:281]: As you know it currently displays the link name to the user here:

Screenshot 2023-02-13 at 22-37-32 sdffsf

I don't want to change that. Using `{docs,.github,.gitea}/README.md` is already an edge case and it is easier for users to understand the special cases if they are a limited number.

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
Please update branch

**Comment by kousu** [comment]
Thank you @lunny, @delvh and @zeripath :)

**Comment by jolheiser** [comment]
@kousu
Please update again, I've added a label so we can get it merged hopefully without bugging you too many more times. 🙂

**Comment by kousu** [comment]
Of course! Here it comes around again.

I really appreciate getting this in. I've had my eye on this patch for a couple of months now.

Did I do something wrong btw? Did I miss clicking "Allow Edits from Maintainers" or something?

**Comment by jolheiser** [comment]
> Did I do something wrong btw? Did I miss clicking "Allow Edits from Maintainers" or something?

Nope, not your fault, I believe it's a GitHub thing that doesn't let us update it because it's from an org vs a user account.

**Pull request merged**